### ISSUE TITLE
Cannot open files dialog

### ISSUE THREAD

**Issue by hunkydoryrepair** [MicrosoftEdge/WebView2Feedback#728]
In searching for a way to access and upload files through WebView2 I've hit a roadblock. 
Note, drag-drop of files into webview2 is working great. I can get a filelist this way. 
But I have no complete solution for opening a file dialog and retrieving files. Without getting files into the File control, I cannot open them in javascript. I'm not sure if I'm running into a security/permissions problem but it doesn't seem like it.

Running on webforms with the GA client 1.0.664.37

Issues #523 addressed this, and it **almost** works. Here's what I do:

- Monitor clicks on input[type=file] elements in javascript
- Send a WebMessage to host application, including generated selector for the element clicked.
- Open file dialog, receiving array of file names
- Using DevTools Protocol, find the nodeId using the selector passed in WebMessage
- Using that nodeId, call DevTools "DOM.setFileInputFiles", including array of filenames.

Everything works up until the final step. My call to DOM.setFileInputFiles matches other online examples, but even passing an empty array, it always throws an ArgumentException: "Value does not fall within the expected range". It occurs in `System.Runtime.InteropServices.Marshal.ThrowExceptionForHR` and the HRESULT in exception is -2147024809, which is just "parameter error".

the actual call looks like:
```
await cwv2.CallDevToolsProtocolMethodAsync(
    "DOM.setFileInputFiles", 
    "{\"nodeId\":174,\"files\":[\"C:/Users/garr/Pictures/chair.jpg\"]}");
```

However, I've tried various changes, such as not passing nodeId at all, sending "[]" for `files`, using '\\' vs '/', reordering parameters.

At this point, I really wish there was a protocol for opening file dialog. I don't mind implementing the dialog in code, but using the DevTools api seems like a real hack, especially when there are multiple calls involved to get a nodeId and I have to construct a selector, assuming either a `name` or `id` attribute exists.. The fact that it doesn't work makes it less than a hack; it is a non-solution.

I am able to use other DevTools apis, even on the same `nodeId`. I can change visibility and other attributes. But setting the files always fails.  For reference, here is the code that does not work, failure is on the DOM.setFileInputFiles call:

```
        private async void ReturnFiles(String[] files, String id)
        {
            try
            {
                JavaScriptSerializer jss = new JavaScriptSerializer();
                var cwv2 = this.ctlWebBrowser.CoreWebView2;
                // get document root
                var node = await cwv2.CallDevToolsProtocolMethodAsync("DOM.getDocument", "{}");
                var nodeinfo = jss.Deserialize<Dictionary<String,object>>(node);
                if (nodeinfo.ContainsKey("root") &&
                    nodeinfo["root"] is Dictionary<string, object>)
                {
                    var query =
                    jss.Serialize(new
                    {
                        nodeId = ((Dictionary<string, object>)nodeinfo["root"])["nodeId"],
                        selector = id
                    });
                    node = await cwv2.CallDevToolsProtocolMethodAsync("DOM.querySelector", query);
                    if (!String.IsNullOrEmpty(node))
                    {
                        nodeinfo = jss.Deserialize<Dictionary<String, object>>(node);

                        var parameters = jss.Serialize(new
                        {
                            nodeId = nodeinfo["nodeId"],
                            files = files.Select(s => s.Replace('\\', '/'))
                        });
                        // we make it all the way to final step here!
                        await cwv2.CallDevToolsProtocolMethodAsync("DOM.setFileInputFiles", parameters);
                    }
                }
            }
            catch(Exception e)
            {
                LogMessage?.Invoke(String.Format("Exception DevTools {0}", e.Message));
            }
        }
```

**Comment by hunkydoryrepair** [comment]
`setFileInputFiles` is part of DevTools 1.3, but I have verified other 1.3 endpoints are available.  DOM.describeNode works just fine and can also be used to verify the nodeId I receive is for an INPUT of type "file".

**Comment by ukandrewc** [comment]
If you just want to open the file dialog from code then: `myFileInput.click()`

**Comment by hunkydoryrepair** [comment]
That's what I do and it works in Chrome and Edge standalone, but in WebView2 nothing shows

**Comment by hunkydoryrepair** [comment]
also tested in latest prerelease 1.0.721

**Comment by ukandrewc** [comment]
Working here, in DevTools and javascript:

![image](https://user-images.githubusercontent.com/30246320/101818862-62582900-3b1c-11eb-9a3a-54ed43b278ca.png)

**Comment by hunkydoryrepair** [comment]
and what is that, a sample app? Was anything special done to accomplish the File Dialog coming up? Which version?

I do think it used to work, but not working with latest.

Running WebView2 control in WinForms, even actually clicking on the "Choose File" button does nothing. I get the briefest glimpse of a change to the mouse pointer, like loading a page, but there are no navigation events firing.

**Comment by ukandrewc** [comment]
It's a screenshot of a Brontosaurus on a skateboard??

**Comment by hunkydoryrepair** [comment]
That's not super helpful. It doesn't appear to be any of the sample apps for WebView2 so I cannot check what version or code it uses. But, to confirm, I downloaded and built the sample app for WinForms, went to Google image search, clicked "browse" to open file dialog and....it doesn't do anything.  Tried several other pages and, sure enough, no file dialog.

**Comment by ukandrewc** [comment]
Just a plain WebView2 V1.0.707.0, you can see that it's just using **$0.click()**

**Comment by hunkydoryrepair** [comment]
Ok, it turns out the problem with the file dialog is related to running the application in elevated mode.  I have noticed before that if the WinForms process is elevated the msedge process still is not elevated. It must be attempting to open a dialog in the application process or has to do with the parent window being elevated.
I don't think the problem with setFileInputFiles is related to elevation mode, however, so if we need to do a workaround so we can run our process as elevated, we don't have one. But this is less urgent now

**Comment by ukandrewc** [comment]
@hunkydoryrepair No problem at all, it's nice that you said thank you (oh wait a minute, you didn't)

**Comment by mrsteamfist** [comment]
There is someone who asked the same question on open file here #734 . I am closing this since the other issue has gone away.

**Comment by darkguy2008** [comment]
Well, in my case I'm experiencing the same issue, albeit not with running as admin. In my case, I already have the filename set (so I don't want the file dialog to open, it's a temp file that my app created) and I'm experiencing the same `"Value does not fall within the expected range."` error. 

It would be **super duper helpful** if WebView2 would just throw (if any) the raw error from CDP, or at least, allow us to provide a value that **falls within the expected range** that we obviously can't see.

Any pointers?

**Comment by darbid** [comment]
Pretty sure you cannot click the element from JS. Also pretty sure that this is buried deep in CEF and not something Webview2 could work around. Happy to be wrong on this and would love to hear that this security measure can be relaxed.

I recently even looked for a command line arg https://stackoverflow.com/questions/66220472/looking-for-a-chromium-command-line-switch-to-relax-htmlinputelement-restriction in this SO question.

Yes ukandrewc - you can execute a JS click from Dev tools and it works. But not from code running in the browser.

**Comment by darkguy2008** [comment]
> Pretty sure you cannot click the element from JS. Also pretty sure that this is buried deep in CEF and not something Webview2 could work around. Happy to be wrong on this and would love to hear that this security measure can be relaxed.
> 
> I recently even looked for a command line arg https://stackoverflow.com/questions/66220472/looking-for-a-chromium-command-line-switch-to-relax-htmlinputelement-restriction in this SO question.
> 
> Yes ukandrewc - you can execute a JS click from Dev tools and it works. But not from code running in the browser.

You can definitely click the `Input` element to trigger the file upload dialog (I do it in my app), but if you want to set it programatically with WebView2, you can't.

You can also do it through the CDP, so it's not that buried deep into CEF that WebView2 can't do it. For some reason they have a bug where the CDP evaluation is crashing (I've noticed that when you send something to CDP that it doesn't like, for WebView2 it throws the `"Value does not fall within the expected range"` error, even if it doesn't have anything to do with an array or something that would hint this error.

I do this in a previous version of my code that was using Chrome instead of WebView2:
```
await _cdp.Send<object>(_page, "DOM.setFileInputFiles", new { BackendNodeId, files = new string[] { filePath } });
```
See here: https://vanilla.aslushnikov.com/?DOM.setFileInputFiles

And you can also simulate clicks with CDP - just look at Puppeteer's source. You just have to dispatch a mouse event: https://vanilla.aslushnikov.com/?Input.dispatchMouseEvent

**Comment by darbid** [comment]
Here is one alternative for pure JS and interacting with the input element of type file.

SO answer
https://stackoverflow.com/questions/47119426/how-to-set-file-objects-and-length-property-at-filelist-object-where-the-files-a

blog post
https://pqina.nl/blog/the-trouble-with-editing-and-uploading-files-in-the-browser/

Data transfer Ref 
https://html.spec.whatwg.org/multipage/dnd.html#datatransfer


Ps: it's been a long time since I have used Internet Explorer and the Web Browser control for JavaScript injection however from memory both allowed a click() on the input element of type￼=type.

**Comment by champnic** [comment]
FYI - the bug that was preventing `DOM.setFileInputFiles` from working has been fixed in runtimes 100.0.1175.0+. We're continuing to look into the fix for the core issue of not being able to display the dialogs properly. Thanks!

**Comment by yeganehaym** [comment]
we have the same problem
we are running winform app with webview and blazor server app
